# Spotify Top-50-World — Exploratory Data Analysis

This notebook explores the raw Spotify Top-50-World dataset before any 
cleaning or modeling decisions are made. 

Steps:
1. Load the data from a local CSV file
2. Load database credentials securely
3. Upload the raw data to Supabase
4. Explore the data's structure and quality
5. Summarize findings and recommended fixes

## Section 1: Loading the Data

We start by loading the raw CSV exactly as it was published — no cleaning, no
column changes, nothing altered. This mirrors the same ELT philosophy used
elsewhere in this course: land the data untouched first, and do every
transformation later, in a step we can inspect and re-run.

`get_csv_path()` below is a small defensive helper: it keeps prompting until
it receives a path that actually exists and is readable, so a typo doesn't
crash the whole notebook halfway through a live demo.

In [1]:
import os
import pandas as pd

# os lets us check whether a file path exists before trying to open it
# pandas is the standard library for loading and inspecting tabular data

In [2]:
# We'll write a small function that keeps asking the user for a file path
# until it finds a valid, readable file. This avoids crashing the notebook
# on a typo, an empty input, or a file we technically can't access.

def get_csv_path():
    while True:  # keep looping until we get a valid path
        # input() pauses the notebook and waits for you to type something
        path = input("Enter the FULL file path (including filename) of your CSV: ")

        # .strip() removes accidental leading/trailing spaces
        # .strip('"') removes quote marks, in case the path was copied with quotes around it
        path = path.strip().strip('"')

        # Catch the case where the user just pressed Enter with nothing typed
        if path == "":
            print("Error: no input was entered. Please type a file path.\n")
            continue  # skip the rest of this loop iteration, ask again

        # "Check first" step: does a file actually exist at this exact path?
        if not os.path.isfile(path):
            print("Error: no file found at that path. Please check for typos and try again.\n")
            continue

        # "Try, then catch" step: the file exists, but can we actually access it?
        # We read in BINARY mode ("rb") here — not text mode — because we only
        # want to test raw file access, not interpret the bytes as text yet.
        # Reading as text here would risk a UnicodeDecodeError on special
        # characters (accents, non-English names) before we've even reached
        # the real data-loading step. Binary mode avoids that entirely.
        try:
            with open(path, "rb") as f:
                f.read(1)  # try reading just 1 byte, as a lightweight access test
        except PermissionError:
            print("Error: file exists, but you don't have permission to read it.\n")
            continue
        except OSError as e:
            print(f"Error: file exists, but could not be opened. Details: {e}\n")
            continue

        # If we reach this line, the file exists and is readable
        print(f"File found: {path}")
        return path  # exit the function and hand back the valid path

# Call the function — this is what actually triggers the input() prompt
csv_path = get_csv_path()

Error: no input was entered. Please type a file path.

File found: D:\spotify\drive-download-20260806T144220Z-1-001\Spotify_Top_50_World.csv


In [3]:
# Now that we've confirmed the file exists and is readable, load it into memory.
# We specify encoding="utf-8" explicitly, since relying on a system default
# encoding can cause errors on files containing accented or non-English characters.

df = pd.read_csv(csv_path, encoding="utf-8")

print("File loaded successfully.")
print(f"Rows: {df.shape[0]:,} | Columns: {df.shape[1]}")
print(f"Columns found: {list(df.columns)}")

# .head() shows the first 5 rows by default — a quick visual sanity check
df.head()

File loaded successfully.
Rows: 27,800 | Columns: 11
Columns found: ['date', 'position', 'song', 'artist', 'popularity', 'duration_ms', 'album_type', 'total_tracks', 'release_date', 'is_explicit', 'album_cover_url']


,date,position,song,artist,popularity,duration_ms,album_type,total_tracks,release_date,is_explicit,album_cover_url
0,2023-05-18,1,Ella Baila Sola,Eslabon Armado,89,165671,album,16,2023-04-28,False,https://i.scdn.co/image/ab67616d0000b273dfddf1...
1,2023-05-18,2,un x100to,Grupo Frontera & Bad Bunny,99,194563,single,1,2023-04-17,False,https://i.scdn.co/image/ab67616d0000b273716c0b...
2,2023-05-18,3,La Bebe - Remix,Yng Lvcas & Peso Pluma,99,234352,single,2,2023-03-17,True,https://i.scdn.co/image/ab67616d0000b273a04be3...
3,2023-05-18,4,Cupid - Twin Ver.,FIFTY FIFTY,97,174253,single,3,2023-02-24,False,https://i.scdn.co/image/ab67616d0000b27337c0b3...
4,2023-05-18,5,Flowers,Miley Cyrus,91,200600,album,13,2023-03-10,False,https://i.scdn.co/image/ab67616d0000b27358039b...


## Section 2: Connecting to Supabase

Before we can upload anything, we need a live connection to the database.
`get_credentials()` reads a separate JSON file rather than accepting typed
input directly — this keeps the actual password out of the notebook's cell
history and out of anything a screen recording or a shared `.ipynb` file
might expose.

The next cell prints a masked confirmation of what was loaded, so you can
visually confirm the right file was picked up without ever showing the real
password on screen. Only after that do we build the SQLAlchemy engine and
test it with a bare `SELECT 1` — the simplest possible query, since it
doesn't depend on any table existing yet, only on the connection and
credentials being valid.

In [11]:
import json

# These are the fields we expect inside the credentials JSON file
REQUIRED_KEYS = ["db_host", "db_port", "db_name", "db_user", "db_password"]

def get_credentials():
    while True:
        cred_path = input("Enter the FULL file path (including filename) of your credentials JSON file: ")
        cred_path = cred_path.strip().strip('"')

        if cred_path == "":
            print("Error: no input was entered. Please type a file path.\n")
            continue

        if not os.path.isfile(cred_path):
            print("Error: no file found at that path. Please check for typos and try again.\n")
            continue

        # The file exists — now check whether its contents are valid JSON.
        # This is where "check first" (isfile) isn't enough on its own: a file
        # can exist but still contain broken or incomplete data.
        try:
            with open(cred_path, "r", encoding="utf-8") as f:
                creds = json.load(f)
        except json.JSONDecodeError:
            print("Error: file exists, but is not valid JSON. Please check its formatting.\n")
            continue
        except OSError as e:
            print(f"Error: file exists, but could not be opened. Details: {e}\n")
            continue

        # Check that every key we need is actually present in the file
        missing = [k for k in REQUIRED_KEYS if k not in creds]
        if missing:
            print(f"Error: file is valid JSON, but is missing these required keys: {missing}\n")
            continue

        print("Credentials file found, and all required fields are present.")
        return creds

credentials = get_credentials()

Credentials file found, and all required fields are present.


In [12]:
# It's good practice to confirm what was loaded WITHOUT ever printing
# sensitive values in full — useful if this notebook is ever run on a
# shared screen, projector, or recording.

def mask(value, keep=3):
    value = str(value)
    if len(value) <= keep:
        return "*" * len(value)
    # Show only the first few characters, mask the rest
    return value[:keep] + "*" * (len(value) - keep)

print("Connection details loaded (sensitive values masked):")
print(f"  db_host      : {credentials['db_host']}")       # hostname isn't sensitive, safe to show fully
print(f"  db_port      : {credentials['db_port']}")        # not sensitive
print(f"  db_name      : {credentials['db_name']}")        # not sensitive
print(f"  db_user      : {mask(credentials['db_user'])}")  # partially mask, contains project reference
print(f"  db_password  : {mask(credentials['db_password'])}")  # always mask

Connection details loaded (sensitive values masked):
  db_host      : aws-0-ap-northeast-2.pooler.supabase.com
  db_port      : 5432
  db_name      : postgres
  db_user      : pos**************************
  db_password  : Shi******************


In [13]:
# sqlalchemy is the standard library for managing database connections in Python.
# It works with many database types (Postgres, MySQL, SQLite, etc.) through a
# consistent interface, so we don't have to write raw connection code ourselves.
from sqlalchemy import create_engine, text

# Postgres connection strings follow a standard format:
# postgresql+psycopg2://username:password@host:port/database_name
#
# We build this string from the credentials we already loaded, rather than
# typing it manually — this avoids retyping sensitive values anywhere.
connection_string = (
    f"postgresql+psycopg2://{credentials['db_user']}:{credentials['db_password']}"
    f"@{credentials['db_host']}:{credentials['db_port']}/{credentials['db_name']}"
)

# create_engine() doesn't connect immediately — it just prepares the connection
# details. The actual connection attempt happens the first time we use it.
engine = create_engine(connection_string)

# Let's test the connection with the simplest possible query: SELECT 1
# This has no dependency on any table existing yet — it only confirms
# that Python can reach the database and authenticate successfully.
try:
    with engine.connect() as conn:
        result = conn.execute(text("SELECT 1"))
        print("Connection successful. Test query returned:", result.scalar())
except Exception as e:
    print("Connection failed. Details:")
    print(e)

Connection successful. Test query returned: 1


## Section 3: Uploading the Raw Data

With a working connection, we push the DataFrame straight into Postgres
using `to_sql()`, then immediately read it back into `df_db`. From this
point forward, `df_db` — not the original local `df` — is treated as the
single source of truth. This is the industry-standard pattern: once data
lives in a proper database, that becomes the reference copy, not whatever
CSV happens to be sitting on someone's laptop.

In [14]:
# Now that we have a working connection (engine) and our data loaded (df),
# we can push the DataFrame straight into a Postgres table using to_sql().
#
# if_exists="replace" means: if a table with this name already exists,
# drop it and recreate it fresh. This is fine for a first upload, but worth
# being careful with later — it will silently overwrite existing data.

table_name = "spotify_top_50_world"

df.to_sql(
    table_name,
    con=engine,
    if_exists="replace",   # options: "replace", "append", "fail"
    index=False,           # don't upload pandas' internal row index as a column
    chunksize=1000         # send data in batches of 1000 rows, not all at once
)

print(f"Upload complete. Table '{table_name}' created with {df.shape[0]:,} rows.")

Upload complete. Table 'spotify_top_50_world' created with 27,800 rows.


In [15]:
# Now we read the data back FROM Supabase, rather than from the original
# local file. From this point on, our analysis should treat the database
# as the single source of truth — this is the industry-standard pattern:
# once data is loaded into a proper database, that becomes the reference
# copy, not the original CSV sitting on someone's laptop.

query = f"SELECT * FROM {table_name}"
df_db = pd.read_sql(query, con=engine)

print(f"Rows retrieved from database: {df_db.shape[0]:,}")
print(f"Columns retrieved: {list(df_db.columns)}")

# Quick sanity check: does the row count match what we originally uploaded?
if df_db.shape[0] == df.shape[0]:
    print("Row count matches the original upload.")
else:
    print("Warning: row count does NOT match. Investigate before proceeding.")

df_db.head()

Rows retrieved from database: 27,800
Columns retrieved: ['date', 'position', 'song', 'artist', 'popularity', 'duration_ms', 'album_type', 'total_tracks', 'release_date', 'is_explicit', 'album_cover_url']
Row count matches the original upload.


,date,position,song,artist,popularity,duration_ms,album_type,total_tracks,release_date,is_explicit,album_cover_url
0,2023-05-18,1,Ella Baila Sola,Eslabon Armado,89,165671,album,16,2023-04-28,False,https://i.scdn.co/image/ab67616d0000b273dfddf1...
1,2023-05-18,2,un x100to,Grupo Frontera & Bad Bunny,99,194563,single,1,2023-04-17,False,https://i.scdn.co/image/ab67616d0000b273716c0b...
2,2023-05-18,3,La Bebe - Remix,Yng Lvcas & Peso Pluma,99,234352,single,2,2023-03-17,True,https://i.scdn.co/image/ab67616d0000b273a04be3...
3,2023-05-18,4,Cupid - Twin Ver.,FIFTY FIFTY,97,174253,single,3,2023-02-24,False,https://i.scdn.co/image/ab67616d0000b27337c0b3...
4,2023-05-18,5,Flowers,Miley Cyrus,91,200600,album,13,2023-03-10,False,https://i.scdn.co/image/ab67616d0000b27358039b...


## Section 4: Exploring the Data

Now that our data lives in a proper database, let's investigate its structure 
and quality — checking for missing values, duplicates, inconsistent formats, 
and anything else worth understanding before we design a data model.

We'll work from `df_db` (the data read back from Supabase) from here on, 
since that's now our source of truth.

In [16]:
# .info() gives a compact summary: column names, data types, and how many
# non-null (i.e. non-missing) values each column has
df_db.info()

<class 'pandas.DataFrame'>
RangeIndex: 27800 entries, 0 to 27799
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   date             27800 non-null  str  
 1   position         27800 non-null  int64
 2   song             27800 non-null  str  
 3   artist           27800 non-null  str  
 4   popularity       27800 non-null  int64
 5   duration_ms      27800 non-null  int64
 6   album_type       27800 non-null  str  
 7   total_tracks     27800 non-null  int64
 8   release_date     27800 non-null  str  
 9   is_explicit      27800 non-null  bool 
 10  album_cover_url  27800 non-null  str  
dtypes: bool(1), int64(4), str(6)
memory usage: 2.1 MB


In [17]:
# A more explicit missing-values check: count how many nulls exist per column,
# and what percentage of the total that represents
missing_counts = df_db.isnull().sum()
missing_pct = (missing_counts / len(df_db)) * 100

missing_summary = pd.DataFrame({
    "missing_count": missing_counts,
    "missing_pct": missing_pct.round(2)
})

# Only show columns that actually have at least one missing value
missing_summary[missing_summary["missing_count"] > 0]

,missing_count,missing_pct


In [18]:
# A duplicate row here would mean the exact same song, on the exact same day,
# at the exact same position, with identical popularity/duration/etc. — 
# that would suggest something went wrong in either the source data or upload.

exact_duplicates = df_db.duplicated().sum()
print(f"Fully duplicate rows (every column identical): {exact_duplicates}")

# A more meaningful check: are there duplicate (date, position) pairs?
# Logically, only ONE song can occupy a given position on a given day —
# if this shows any count above 1, that's a genuine data integrity problem.
duplicate_date_position = df_db.duplicated(subset=["date", "position"]).sum()
print(f"Duplicate (date, position) pairs: {duplicate_date_position}")

Fully duplicate rows (every column identical): 0
Duplicate (date, position) pairs: 0


In [19]:
# Let's check: is a song title alone enough to uniquely identify a track,
# or can the same title belong to different artists?

distinct_songs = df_db["song"].nunique()
distinct_song_artist_pairs = df_db[["song", "artist"]].drop_duplicates().shape[0]

print(f"Distinct song titles: {distinct_songs}")
print(f"Distinct (song, artist) combinations: {distinct_song_artist_pairs}")

if distinct_songs != distinct_song_artist_pairs:
    print("\nThese numbers don't match — meaning at least one song title")
    print("is shared by more than one artist. Let's find out which ones.")

    # Group by song title, and count how many DIFFERENT artists share that title
    song_artist_counts = df_db.groupby("song")["artist"].nunique()

    # Keep only the songs where more than 1 distinct artist is attached
    conflicting_titles = song_artist_counts[song_artist_counts > 1]

    print(f"\nNumber of song titles shared across multiple artists: {len(conflicting_titles)}")
    conflicting_titles.head(10)

Distinct song titles: 794
Distinct (song, artist) combinations: 827

These numbers don't match — meaning at least one song title
is shared by more than one artist. Let's find out which ones.

Number of song titles shared across multiple artists: 31


In [20]:
# release_date is currently stored as an "object" (text) column, not a proper
# date type. Let's investigate WHY — by checking the length and pattern of
# the values, rather than assuming they're all consistent.

# A full date like "2023-04-28" is 10 characters long.
# Let's see if every value in this column actually follows that length.

release_date_lengths = df_db["release_date"].str.len().value_counts().sort_index()
print("Distribution of release_date string lengths:")
print(release_date_lengths)

Distribution of release_date string lengths:
release_date
4       101
7         4
10    27695
Name: count, dtype: int64


In [21]:
# Let's look at actual examples of each length, to understand what format
# each one represents

for length in release_date_lengths.index:
    example = df_db[df_db["release_date"].str.len() == length]["release_date"].iloc[0]
    count = release_date_lengths[length]
    print(f"Length {length}: example = '{example}'  (appears {count} times)")

Length 4: example = '1962'  (appears 101 times)
Length 7: example = '1957-09'  (appears 4 times)
Length 10: example = '2023-04-28'  (appears 27695 times)


In [22]:
# Position should logically only ever be between 1 and 50, since this is a
# "Top 50" chart. Let's verify that assumption rather than trust it blindly.

print("Position range check:")
print(f"  Min position: {df_db['position'].min()}")
print(f"  Max position: {df_db['position'].max()}")

# popularity is documented as a 0-100 score. Let's confirm no values fall outside that.
print("\nPopularity range check:")
print(f"  Min popularity: {df_db['popularity'].min()}")
print(f"  Max popularity: {df_db['popularity'].max()}")

# duration_ms should always be a positive number (a song can't have zero or
# negative length). Let's check for anything suspicious.
print("\nDuration range check:")
print(f"  Min duration_ms: {df_db['duration_ms'].min()}")
print(f"  Max duration_ms: {df_db['duration_ms'].max()}")

Position range check:
  Min position: 1
  Max position: 50

Popularity range check:
  Min popularity: 0
  Max popularity: 100

Duration range check:
  Min duration_ms: 41487
  Max duration_ms: 456933


In [23]:
# .describe() gives a quick statistical overview of all numeric columns at once:
# count, mean, standard deviation, min, max, and quartiles (25%/50%/75%)
df_db.describe()

,position,popularity,duration_ms,total_tracks
count,27800.000000,27800.000000,27800.000000,27800.000000
mean,25.500000,89.621331,196898.287338,10.160468
std,14.431129,9.969284,41821.134307,8.282919
min,1.000000,0.000000,41487.000000,1.000000
25%,13.000000,88.000000,171291.000000,1.000000
50%,25.500000,91.000000,191700.000000,11.000000
75%,38.000000,94.000000,222369.000000,15.000000
max,50.000000,100.000000,456933.000000,119.000000


In [24]:
# Let's look directly at the specific rows behind each suspicious extreme,
# rather than just noting the number and moving on.

print("Rows with popularity == 0:")
display(df_db[df_db["popularity"] == 0][["song", "artist", "date", "popularity"]])

print("\nRows with the shortest duration:")
display(df_db[df_db["duration_ms"] == df_db["duration_ms"].min()][["song", "artist", "duration_ms"]])

print("\nRows with the highest total_tracks:")
display(df_db[df_db["total_tracks"] == df_db["total_tracks"].max()][["song", "artist", "album_type", "total_tracks"]].drop_duplicates())

Rows with popularity == 0:


,song,artist,date,popularity
2205,LADY GAGA,Peso Pluma,2023-07-01,0
2216,LUNA,Peso Pluma,2023-07-01,0
2600,Mine (Taylor's Version),Taylor Swift,2023-07-09,0
2601,Back To December (Taylor's Version),Taylor Swift,2023-07-09,0
2602,Enchanted (Taylor's Version),Taylor Swift,2023-07-09,0
...,...,...,...,...
22069,FIELD TRIP,¥$ & Kanye West & Ty Dolla $ign,2024-08-04,0
22083,SLIDE,¥$ & Kanye West & Ty Dolla $ign,2024-08-04,0
22653,New Woman (feat. ROSALÍA),LISA & ROSALÍA,2024-08-17,0
22659,Die With A Smile,Lady Gaga & Bruno Mars,2024-08-17,0



Rows with the shortest duration:


,song,artist,duration_ms
21016,Trouble,Eminem,41487



Rows with the highest total_tracks:


,song,artist,album_type,total_tracks
9986,Happy Xmas (War Is Over) - Remastered 2010,John Lennon,compilation,119


In [26]:
import re

# We're looking for characters that shouldn't normally appear in song/artist
# names — things outside standard printable text, which often indicate
# encoding corruption (like the $ symbol turning into ¥ or ¤).

def find_unusual_chars(series):
    unusual_rows = []
    for value in series.dropna().unique():
        # Flag any character that isn't a standard letter, number, space,
        # or common punctuation
        if re.search(r"[^\w\s\-\'\".,()&!?:/]", value):
            unusual_rows.append(value)
    return unusual_rows

unusual_artists = find_unusual_chars(df_db["artist"])
unusual_songs = find_unusual_chars(df_db["song"])

print(f"Artist names with unusual characters: {len(unusual_artists)}")
for name in unusual_artists[:15]:
    print(f"  {name}")

print(f"\nSong titles with unusual characters: {len(unusual_songs)}")
for title in unusual_songs[:15]:
    print(f"  {title}")

Artist names with unusual characters: 4
  ¥$ & Kanye West & Ty Dolla $ign
  Kanye West & Ty Dolla $ign
  *NSYNC
  Earth, Wind & Fire & sped up + slowed

Song titles with unusual characters: 28
  Am I Dreaming (Metro Boomin & A$AP Rocky, Roisee)
  Barbie World (with Aqua) [From Barbie The Album]
  Sparks Fly (Taylor’s Version)
  I Can See You (Taylor’s Version) (From The Vault)
  Electric Touch (feat. Fall Out Boy) (Taylor’s Version) (From The Vault)
  When Emma Falls in Love (Taylor’s Version) (From The Vault)
  Castles Crumbling (feat. Hayley Williams) (Taylor’s Version) (From The Vault)
  Foolish One (Taylor’s Version) (From The Vault)
  Timeless (Taylor’s Version) (From The Vault)
  Ours (Taylor’s Version)
  Superman (Taylor’s Version)
  What Was I Made For? [From The Motion Picture "Barbie"]
  pretty isn’t pretty
  Cruel Summer - Live from TS | The Eras Tour
  This Love (Taylor’s Version)


In [27]:
# Let's search specifically for any artist value containing "Dolla" — 
# this will show us every variant of this artist credit that exists,
# clean or corrupted, side by side.

dolla_variants = df_db[df_db["artist"].str.contains("Dolla", na=False)]["artist"].value_counts()
print("All variants of artist names containing 'Dolla':")
print(dolla_variants)

All variants of artist names containing 'Dolla':
artist
¥$ & Kanye West & Ty Dolla $ign    120
Kanye West & Ty Dolla $ign          40
Name: count, dtype: int64


In [28]:
# Now let's search specifically for mojibake-style characters — these are
# the classic tell-tale symbols (¥, ¤, Â, Ã, etc.) that appear when UTF-8
# bytes get misread using the wrong encoding. This is a more precise check
# than our earlier broad punctuation filter.

mojibake_pattern = r"[¥¤ÂÃ]"

corrupted_artists = df_db[df_db["artist"].str.contains(mojibake_pattern, na=False, regex=True)]
print(f"\nRows with likely mojibake corruption in artist column: {len(corrupted_artists)}")
corrupted_artists[["song", "artist", "date"]].drop_duplicates()


Rows with likely mojibake corruption in artist column: 120


,song,artist,date
9495,Vultures,¥$ & Kanye West & Ty Dolla $ign,2023-11-23
13700,CARNIVAL,¥$ & Kanye West & Ty Dolla $ign,2024-02-16
13713,BURN,¥$ & Kanye West & Ty Dolla $ign,2024-02-16
13714,FUK SUMN,¥$ & Kanye West & Ty Dolla $ign,2024-02-16
13721,BACK TO ME,¥$ & Kanye West & Ty Dolla $ign,2024-02-16
...,...,...,...
22174,FIELD TRIP,¥$ & Kanye West & Ty Dolla $ign,2024-08-06
22230,FIELD TRIP,¥$ & Kanye West & Ty Dolla $ign,2024-08-07
22287,FIELD TRIP,¥$ & Kanye West & Ty Dolla $ign,2024-08-08
22342,FIELD TRIP,¥$ & Kanye West & Ty Dolla $ign,2024-08-09


In [29]:
# Same mojibake check, applied to song titles instead of artist names
corrupted_songs = df_db[df_db["song"].str.contains(mojibake_pattern, na=False, regex=True)]
print(f"Rows with likely mojibake corruption in song column: {len(corrupted_songs)}")
corrupted_songs[["song", "artist", "date"]].drop_duplicates()

Rows with likely mojibake corruption in song column: 6


,song,artist,date
18645,MTG QUEM NÃO QUER SOU EU,DJ TOPO & Seu Jorge & Mc Leozin,2024-05-26
18694,MTG QUEM NÃO QUER SOU EU,DJ TOPO & Seu Jorge & Mc Leozin,2024-05-27
18898,MTG QUEM NÃO QUER SOU EU,DJ TOPO & Seu Jorge & Mc Leozin,2024-05-31
18947,MTG QUEM NÃO QUER SOU EU,DJ TOPO & Seu Jorge & Mc Leozin,2024-06-01
18996,MTG QUEM NÃO QUER SOU EU,Various Artists,2024-06-02
19040,MTG QUEM NÃO QUER SOU EU,Various Artists,2024-06-03


In [30]:
# Refined pattern: ¥ and ¤ are genuinely unusual in any language's normal text
# (currency/special symbols with no place in a song or artist name).
# Ã and Â, however, are legitimate letters in Portuguese, French, and other
# languages — so we remove them from this check to avoid flagging valid
# international text as if it were broken.

true_mojibake_pattern = r"[¥¤]"

corrupted_songs = df_db[df_db["song"].str.contains(true_mojibake_pattern, na=False, regex=True)]
print(f"Rows with likely genuine mojibake corruption in song column: {len(corrupted_songs)}")
corrupted_songs[["song", "artist", "date"]].drop_duplicates()

Rows with likely genuine mojibake corruption in song column: 0


,song,artist,date


## Section 4 Summary: What We Found

Through direct exploration of the data (not assumptions made in advance), 
we identified the following:

**Structural integrity — clean:**
- No missing values in any of the 11 columns
- No fully duplicate rows
- No duplicate (date, position) pairs — the "one song per position per day" 
  rule holds true across all 27,800 rows
- `position` stays within 1–50, `popularity` stays within 0–100, as expected

**Issues identified:**

1. **`release_date` has inconsistent precision** — 27,695 full dates (YYYY-MM-DD), 
   101 year-only values (e.g. "1962"), and 4 year-month values (e.g. "1957-09"). 
   This column is currently stored as text, not a proper date type, and would 
   break date-based calculations (like days-since-release) without cleaning.

2. **`song` alone is not a safe identifier** — 794 distinct song titles vs. 827 
   distinct (song, artist) combinations. 31 song titles are shared across 
   multiple different artists.

3. **Artist-name encoding corruption** — the collaboration "Kanye West & Ty 
   Dolla $ign" appears as two different values: a corrupted version 
   ("¥$ & Kanye West & Ty Dolla $ign", 120 rows) and a clean version 
   ("Kanye West & Ty Dolla $ign", 40 rows). This silently fragments one 
   artist's chart history into two separate identities.

4. **`popularity == 0` appears 182 times** across otherwise normal-looking 
   rows from well-known artists — plausible but not yet explained; worth 
   flagging rather than deleting or assuming it's an error.

5. **Attribution inconsistency** — the song "MTG QUEM NÃO QUER SOU EU" is 
   credited to different artist values ("DJ TOPO & Seu Jorge & Mc Leozin" vs. 
   "Various Artists") across different dates. A separate issue from the 
   encoding corruption above — likely a genuine re-crediting or inconsistent 
   logging over time, not yet resolved.

**Methodology caution (kept deliberately, for its teaching value):** an early, 
overly broad "unusual character" regex scan produced false positives — 
flagging legitimate punctuation (`*`, `[`, `]`, `+`) and legitimate 
non-English letters (`Ã`, `Â`, valid in Portuguese and French) as if they 
were corruption. Narrowing the pattern to only the true mojibake signatures 
(`¥`, `¤`) eliminated the false positives while still catching the real 
issue. Lesson: overly broad anomaly detection manufactures problems that 
don't exist, and treating "not English" as "probably broken" is a bias 
worth guarding against.

## Section 5: Recommendations

Having explored the data directly rather than assumed its structure, we now 
have evidence-backed reasons for each change below. Each recommendation is 
tied directly back to a specific finding from Section 4 — nothing here is 
arbitrary.

### Recommendation 1: Clean and standardize `release_date`

**Finding it addresses:** 101 year-only values, 4 year-month values, stored 
as inconsistent text — would break any date arithmetic (e.g. days-since-release) 
without fixing first.

**Approach:** pad incomplete dates to their most reasonable midpoint — a 
year-only value becomes June 30 of that year; a year-month value becomes 
the 15th of that month. This is an explicit, documented assumption — not 
a claim of precision we don't actually have.

In [31]:
from datetime import datetime

def clean_release_date(value):
    """
    Standardizes release_date values of varying precision into a full date.
    - Full date (YYYY-MM-DD): returned as-is
    - Year-month (YYYY-MM): padded to the 15th of that month
    - Year only (YYYY): padded to June 30 of that year
    """
    value = str(value).strip()

    if len(value) == 10:  # already a full date
        return value
    elif len(value) == 7:  # year-month, e.g. "1957-09"
        return f"{value}-15"
    elif len(value) == 4:  # year only, e.g. "1962"
        return f"{value}-06-30"
    else:
        # Anything else is unexpected — flag it rather than guess
        return None

# Apply the function to create a new, cleaned column.
# We keep the original release_date untouched, so nothing is lost or hidden.
df_db["release_date_cleaned"] = df_db["release_date"].apply(clean_release_date)

# Confirm: did anything fail to convert (return None)?
failed = df_db[df_db["release_date_cleaned"].isnull()]
print(f"Rows that failed to clean: {len(failed)}")

# Now convert the cleaned column to an actual datetime type
df_db["release_date_cleaned"] = pd.to_datetime(df_db["release_date_cleaned"])
print("release_date_cleaned dtype:", df_db["release_date_cleaned"].dtype)

df_db[["release_date", "release_date_cleaned"]].drop_duplicates().head(10)

Rows that failed to clean: 0
release_date_cleaned dtype: datetime64[us]


,release_date,release_date_cleaned
0,2023-04-28,2023-04-28
1,2023-04-17,2023-04-17
2,2023-03-17,2023-03-17
3,2023-02-24,2023-02-24
4,2023-03-10,2023-03-10
5,2023-04-14,2023-04-14
6,2022-12-08,2022-12-08
7,2023-02-25,2023-02-25
8,2022-05-20,2022-05-20
10,2023-01-17,2023-01-17


In [32]:
# Let's specifically check the year-only and year-month cases we found earlier,
# to confirm the padding logic worked as intended — not just that it ran
# without error.

padded_examples = df_db[df_db["release_date"].str.len() < 10][
    ["release_date", "release_date_cleaned"]
].drop_duplicates()

print(f"Number of distinct padded values: {len(padded_examples)}")
padded_examples.head(10)

Number of distinct padded values: 11


,release_date,release_date_cleaned
8365,1962,1962-06-30
9584,1963,1963-06-30
11034,1957-09,1957-09-15
11040,2002,2002-06-30
11094,1981,1981-06-30
11141,1971,1971-06-30
11147,1947,1947-06-30
12484,1999,1999-06-30
22449,1984,1984-06-30
23494,1998,1998-06-30


### Recommendation 2: Fix artist-name encoding corruption

**Finding it addresses:** the collaboration credit "Kanye West & Ty Dolla 
$ign" was silently split into two different artist values by an encoding 
bug — a corrupted variant (`¥$ & Kanye West & Ty Dolla $ign`, 120 rows) and 
a clean variant (`Kanye West & Ty Dolla $ign`, 40 rows). Left alone, this 
fragments one real artist's chart history into two separate identities in 
any group-by or ranking.

**Approach:** an exact-string replace, not a regex pattern. The corrupted 
value was already known precisely from the investigation in Section 4, so a 
targeted replace fixes exactly this one case without risking any unintended 
change to a different row that happens to share some of the same characters.

In [33]:
# We replace the exact corrupted string with its clean equivalent.
# Using the exact string (rather than a broad pattern) avoids accidentally
# altering any other row that happens to contain similar characters.

corrupted_value = "¥$ & Kanye West & Ty Dolla $ign"
clean_value = "Kanye West & Ty Dolla $ign"

rows_before = (df_db["artist"] == corrupted_value).sum()

df_db["artist"] = df_db["artist"].replace(corrupted_value, clean_value)

rows_after_corrupted = (df_db["artist"] == corrupted_value).sum()
rows_after_clean = (df_db["artist"] == clean_value).sum()

print(f"Rows with corrupted value before fix: {rows_before}")
print(f"Rows with corrupted value after fix: {rows_after_corrupted}")
print(f"Rows with clean value after fix: {rows_after_clean}")

Rows with corrupted value before fix: 120
Rows with corrupted value after fix: 0
Rows with clean value after fix: 160


### Recommendation 3: Split the flat table into a proper data model

**Finding it addresses:** `song` alone is not a safe identifier (31 titles 
shared across multiple artists) — so any relationship or lookup must be 
built on the combination of `song` + `artist`, not `song` alone.

**Additional reasoning:** our single flat table currently repeats song-level 
information (artist, release date, album type, total tracks) on every single 
daily row — once per day that song charted. This is redundant, and risks 
inconsistency (as we saw with the Ty Dolla $ign case) if the same song's 
details are ever entered differently across different rows.

**Approach:** build a composite key from `song` + `artist`, then split the 
data into three tables:
- `fact_chart_entries` — one row per song, per day, per position (daily-changing data only)
- `dim_songs` — one row per unique song+artist combination (fixed song-level attributes)
- `dim_dates` — one row per calendar day (calendar attributes)

In [34]:
# song alone isn't a safe identifier (we proved this in Section 4 — 31 titles
# are shared across different artists). So we build a composite key by
# combining song + artist together, separated by a delimiter that's unlikely
# to appear naturally in either field.
#
# We lowercase both parts before combining, rather than using the display
# casing directly. Power BI's own model engine compares text CASE-
# INSENSITIVELY (a documented VertiPaq behavior, unrelated to this dataset) —
# so a song logged as "BYE" on one chart date and "Bye" on another would
# otherwise produce two DIFFERENT song_key values here, but Power BI would
# still treat them as duplicates once loaded, breaking the dim_songs
# relationship. Lowercasing the key up front matches Power BI's own
# comparison rules, so both systems agree on what counts as "the same song."
# Only the key is lowercased — the original song and artist columns (used
# for display) keep their original casing untouched.

df_db["song_key"] = (
    df_db["song"].str.strip().str.lower() + " || " + df_db["artist"].str.strip().str.lower()
)

# Sanity check: this key should now be unique per (song, artist) pair, using
# the SAME case-insensitive comparison as song_key itself — otherwise this
# check would wrongly flag a "mismatch" for the exact case-variant pairs we
# just deliberately collapsed together (e.g. BYE/Bye, VULTURES/Vultures).
distinct_keys = df_db["song_key"].nunique()
distinct_song_artist_pairs = (
    df_db["song"].str.strip().str.lower() + " || " + df_db["artist"].str.strip().str.lower()
).nunique()

print(f"Distinct song_key values: {distinct_keys}")
print(f"Distinct (song, artist) pairs (case-insensitive): {distinct_song_artist_pairs}")

if distinct_keys == distinct_song_artist_pairs:
    print("song_key is a valid, unique identifier for each song+artist combination.")
else:
    print("Mismatch — investigate before proceeding.")

Distinct song_key values: 817
Distinct (song, artist) pairs (case-insensitive): 817
song_key is a valid, unique identifier for each song+artist combination.


In [35]:
# dim_songs should have exactly one row per unique song+artist combination,
# holding only the attributes that stay fixed for that song — not the
# daily-changing ones (date, position, popularity).

song_level_columns = [
    "song_key", "song", "artist", "release_date_cleaned",
    "album_type", "total_tracks", "is_explicit", "album_cover_url", "duration_ms"
]

dim_songs = df_db[song_level_columns].drop_duplicates(subset="song_key").reset_index(drop=True)

print(f"dim_songs row count: {dim_songs.shape[0]}")
dim_songs.head()

dim_songs row count: 817


,song_key,song,artist,release_date_cleaned,album_type,total_tracks,is_explicit,album_cover_url,duration_ms
0,ella baila sola || eslabon armado,Ella Baila Sola,Eslabon Armado,2023-04-28,album,16,False,https://i.scdn.co/image/ab67616d0000b273dfddf1...,165671
1,un x100to || grupo frontera & bad bunny,un x100to,Grupo Frontera & Bad Bunny,2023-04-17,single,1,False,https://i.scdn.co/image/ab67616d0000b273716c0b...,194563
2,la bebe - remix || yng lvcas & peso pluma,La Bebe - Remix,Yng Lvcas & Peso Pluma,2023-03-17,single,2,True,https://i.scdn.co/image/ab67616d0000b273a04be3...,234352
3,cupid - twin ver. || fifty fifty,Cupid - Twin Ver.,FIFTY FIFTY,2023-02-24,single,3,False,https://i.scdn.co/image/ab67616d0000b27337c0b3...,174253
4,flowers || miley cyrus,Flowers,Miley Cyrus,2023-03-10,album,13,False,https://i.scdn.co/image/ab67616d0000b27358039b...,200600


### Building `dim_dates`

We generate this calendar deliberately using `pd.date_range()` rather than
pulling the unique values out of `df_db["date"]`. That distinction matters:
if the source data were ever missing a snapshot day (a scraper hiccup, a
gap in collection), pulling unique dates from the fact data would silently
carry that gap into the calendar too. A proper "Date table" in Power BI
should never have gaps — month-over-month and running-total measures rely
on every single day being present, whether or not a chart snapshot happened
that day. Generating the range explicitly from min to max guarantees that.

In [36]:
# dim_dates: one row per calendar day, independent of whether that day
# actually appears in the fact data. This is what makes it safe to use
# as an official Power BI "Date Table" later.

date_min = pd.to_datetime(df_db["date"]).min()
date_max = pd.to_datetime(df_db["date"]).max()

dim_dates = pd.DataFrame({
    "date": pd.date_range(start=date_min, end=date_max, freq="D")
})

# Calculated columns below all describe the date itself (not the song or
# the chart entry), so per our placement rule, they belong on this table.
dim_dates["year"] = dim_dates["date"].dt.year
dim_dates["month_num"] = dim_dates["date"].dt.month
dim_dates["month_name"] = dim_dates["date"].dt.strftime("%B")
dim_dates["year_month"] = dim_dates["date"].dt.strftime("%Y-%m")

# weekday_num uses the Monday=1 ... Sunday=7 convention to match DAX's
# WEEKDAY(date, 2). Pandas' native dt.dayofweek is Monday=0, so we add 1
# to line the two systems up exactly.
dim_dates["weekday_num"] = dim_dates["date"].dt.dayofweek + 1
dim_dates["weekday_name"] = dim_dates["date"].dt.strftime("%A")

print(f"dim_dates row count: {dim_dates.shape[0]}")
print(f"Date range: {dim_dates['date'].min().date()} to {dim_dates['date'].max().date()}")
dim_dates.head()

dim_dates row count: 560
Date range: 2023-05-18 to 2024-11-27


,date,year,month_num,month_name,year_month,weekday_num,weekday_name
0,2023-05-18,2023,5,May,2023-05,4,Thursday
1,2023-05-19,2023,5,May,2023-05,5,Friday
2,2023-05-20,2023,5,May,2023-05,6,Saturday
3,2023-05-21,2023,5,May,2023-05,7,Sunday
4,2023-05-22,2023,5,May,2023-05,1,Monday


Let's confirm the calendar we just built actually matches the real chart 
data — comparing `dim_dates` (every possible day) against the dates that 
actually appear in `df_db` (days that were actually snapshotted) tells us 
whether any calendar days are silently missing.

In [37]:
# dim_dates is the full calendar (560 days). df_db["date"] is what actually
# has chart snapshots. The difference between them tells us exactly which
# calendar days have no data — information we'd lose if we built the
# calendar from df_db["date"] directly instead of from date_range().

source_dates = set(pd.to_datetime(df_db["date"]).dt.date)
calendar_dates = set(dim_dates["date"].dt.date)

missing_dates = sorted(calendar_dates - source_dates)

print(f"Calendar days: {len(calendar_dates)}")
print(f"Source snapshot days: {len(source_dates)}")
print(f"Missing days: {len(missing_dates)}")
missing_dates

Calendar days: 560
Source snapshot days: 556
Missing days: 4


[datetime.date(2024, 3, 14),
 datetime.date(2024, 3, 25),
 datetime.date(2024, 7, 11),
 datetime.date(2024, 8, 13)]

### Finding 6: Calendar has 4 missing snapshot days (revises earlier "verified 
daily" note)

Section 1 confirmed weekday balance (each of the 7 weekdays appears roughly
78-80 times), which was taken as evidence of consistent daily collection.
That check is necessary but not sufficient — it can't detect a small number
of missing individual days, since losing 4 days spread across different
weekdays barely shifts the weekday balance at all.

Comparing the full 560-day calendar (dim_dates, built independently via
pd.date_range) against the 556 actual snapshot days in df_db["date"]
surfaces exactly this: 2024-03-14, 2024-03-25, 2024-07-11, and 2024-08-13
have no chart data. The four dates are scattered (no consecutive run, no
shared weekday, not clustered around a holiday), consistent with isolated
collection failures on those specific days rather than a systematic gap.

Resolution: no action needed on dim_dates itself — it stays a full 560-day
gapless calendar, which is the correct behavior for a Power BI Date Table.
fact_chart_entries will simply have zero rows for these 4 dates, which is
expected and not an error. Flagging this explicitly matters for any
day-over-day or rolling measures later, since a naive DATEADD-based
comparison spanning one of these 4 dates will compare against a day that
genuinely has no data, not a data error.

In [38]:
# Sanity check the missing dates aren't secretly clustered in a way the
# eyeball missed above (e.g. same weekday, same day-of-month).

for d in missing_dates:
    print(f"{d}  ({d.strftime('%A')})")

2024-03-14  (Thursday)
2024-03-25  (Monday)
2024-07-11  (Thursday)
2024-08-13  (Tuesday)


### Verifying the model and building the fact table

Before trusting this model in Power BI, we check both relationships hold
with zero orphans — every song_key referenced in the daily data must exist
in dim_songs, and every date must exist in dim_dates. Given how dim_songs
was built (drop_duplicates on song_key straight from df_db) and dim_dates
was built (a full calendar spanning min to max), both should already be
guaranteed true by construction — but we verify rather than assume, the
same discipline we've used throughout.

Then we build fact_chart_entries: the daily-changing grain only. Song-level
attributes (release date, album type, total tracks, is_explicit, cover art,
duration) now live in dim_songs and would just be redundant repetition here
— that redundancy is exactly what we're removing by normalizing.

In [39]:
# --- Step 1: verify no orphaned foreign keys ---

orphan_song_keys = set(df_db["song_key"]) - set(dim_songs["song_key"])
orphan_dates = set(pd.to_datetime(df_db["date"]).dt.date) - set(dim_dates["date"].dt.date)

print(f"Orphaned song_keys: {len(orphan_song_keys)}")
print(f"Orphaned dates: {len(orphan_dates)}")

if orphan_song_keys or orphan_dates:
    print("STOP — do not proceed until these are resolved.")
else:
    print("Model verified: every fact row has a valid home in both dimension tables.")

# --- Step 2: build fact_chart_entries ---
# Grain: one song, one day, one position. We drop the columns that now
# live in dim_songs (they'd be pure redundant repetition here), and keep
# song/artist as optional read-friendly text alongside the song_key.

fact_chart_entries = df_db[[
    "date", "song_key", "song", "artist", "position", "popularity"
]].copy()

print(f"\nfact_chart_entries row count: {fact_chart_entries.shape[0]}")
print(f"Expected: {df_db.shape[0]} (should match — this is a column drop, not a row filter)")
fact_chart_entries.head()

Orphaned song_keys: 0
Orphaned dates: 0
Model verified: every fact row has a valid home in both dimension tables.

fact_chart_entries row count: 27800
Expected: 27800 (should match — this is a column drop, not a row filter)


,date,song_key,song,artist,position,popularity
0,2023-05-18,ella baila sola || eslabon armado,Ella Baila Sola,Eslabon Armado,1,89
1,2023-05-18,un x100to || grupo frontera & bad bunny,un x100to,Grupo Frontera & Bad Bunny,2,99
2,2023-05-18,la bebe - remix || yng lvcas & peso pluma,La Bebe - Remix,Yng Lvcas & Peso Pluma,3,99
3,2023-05-18,cupid - twin ver. || fifty fifty,Cupid - Twin Ver.,FIFTY FIFTY,4,97
4,2023-05-18,flowers || miley cyrus,Flowers,Miley Cyrus,5,91


### Adding Calculated Columns to `fact_chart_entries`

Two values we need for the upcoming Power BI analyses depend on combining a
dimension-level attribute (a song's release date, which lives in `dim_songs`)
with a fact-level attribute (the specific date a chart entry happened, which
lives here). That combination is exactly the case our placement rule from
Recommendation 3 calls out — it belongs on the fact table, not on either
dimension alone.

- **`days_since_release`** — how many days after its release date a song was
  charting on that particular day. Powers the Song Lifecycle & Trends page
  (average days-since-release at charting, fastest climbs to #1).
- **`is_new_entry`** — flags the very first date each song appears in the
  chart. Powers the Seasonality page (new entries by weekday) and the
  Song Lifecycle page (first chart date).

We're computing both here in pandas rather than as DAX calculated columns in
Power BI — same result, just done once at ETL time instead of recalculated
on every model refresh.

In [40]:
# fact_chart_entries needs release_date_cleaned to compute days_since_release,
# but that column lives in dim_songs, not here. We bring it in via a merge on
# song_key, use it, then drop it — dim_songs stays the single source of truth
# for that value, this merge is only a temporary lookup, not a permanent copy.
#
# Safety: if a previous run of this cell left release_date_cleaned sitting on
# fact_chart_entries (e.g. from an earlier attempt that errored out before
# reaching the final drop), remove it first — otherwise merging it back in
# creates a _x/_y naming collision instead of a clean single column.
if "release_date_cleaned" in fact_chart_entries.columns:
    fact_chart_entries = fact_chart_entries.drop(columns=["release_date_cleaned"])

fact_chart_entries = fact_chart_entries.merge(
    dim_songs[["song_key", "release_date_cleaned"]],
    on="song_key",
    how="left"
)

# "date" came back from Supabase as plain text, not a real date type — convert
# it now so we can do actual date arithmetic against release_date_cleaned,
# which was already converted to a real datetime back in Cell 30.
fact_chart_entries["date"] = pd.to_datetime(fact_chart_entries["date"])

# days_since_release: how many days after release this chart entry occurred.
# Can occasionally be negative — a side effect of the padding assumption from
# Recommendation 1 (a year-only release_date gets padded to June 30, which can
# land after a song's actual first chart appearance). Not treated as an error,
# just an inherited limitation of the original data's precision.
fact_chart_entries["days_since_release"] = (
    fact_chart_entries["date"] - fact_chart_entries["release_date_cleaned"]
).dt.days

# is_new_entry: True on the very first date each song_key appears in the
# chart, False on every other day it appears.
first_chart_date = fact_chart_entries.groupby("song_key")["date"].transform("min")
fact_chart_entries["is_new_entry"] = fact_chart_entries["date"] == first_chart_date

# Drop the merged-in helper column now that days_since_release is computed —
# dim_songs remains the only place release_date_cleaned officially lives.
fact_chart_entries = fact_chart_entries.drop(columns=["release_date_cleaned"])

print(f"days_since_release range: {fact_chart_entries['days_since_release'].min()} to {fact_chart_entries['days_since_release'].max()}")
print(f"Rows with negative days_since_release: {(fact_chart_entries['days_since_release'] < 0).sum()}")
print(f"Total new-entry rows (should equal dim_songs row count, 820): {fact_chart_entries['is_new_entry'].sum()}")
fact_chart_entries.head()

days_since_release range: 1 to 29945
Rows with negative days_since_release: 0
Total new-entry rows (should equal dim_songs row count, 820): 817


,date,song_key,song,artist,position,popularity,days_since_release,is_new_entry
0,2023-05-18,ella baila sola || eslabon armado,Ella Baila Sola,Eslabon Armado,1,89,20,True
1,2023-05-18,un x100to || grupo frontera & bad bunny,un x100to,Grupo Frontera & Bad Bunny,2,99,31,True
2,2023-05-18,la bebe - remix || yng lvcas & peso pluma,La Bebe - Remix,Yng Lvcas & Peso Pluma,3,99,62,True
3,2023-05-18,cupid - twin ver. || fifty fifty,Cupid - Twin Ver.,FIFTY FIFTY,4,97,83,True
4,2023-05-18,flowers || miley cyrus,Flowers,Miley Cyrus,5,91,69,True


### Uploading the Final Model to Supabase

With the model verified — `dim_songs`, `dim_dates`, and `fact_chart_entries` 
all built and cross-checked for orphaned keys — we upload all three as 
separate tables, replacing the single flat `spotify_top_50_world` table 
(which stays in place in the database as a useful "before" comparison, 
since we never dropped it).

Each table gets its own dedicated connection via `engine.begin()`, used 
once and closed immediately, rather than sharing one connection across all 
three uploads. This matters in practice: if a single connection is reused 
and one statement on it fails, that connection's transaction becomes 
invalid and every later statement on the same connection inherits the same 
broken state — even for tables that have nothing wrong with them. Isolating 
each upload to its own connection means one failure can't cascade into the 
others. The read-back verification loop afterward follows the same 
discipline established back in Section 3: never trust an upload without 
confirming the row count independently, on the other side of the connection.

In [41]:
import traceback

# Reset the connection pool first — if the earlier run left a connection
# in a broken transaction state, this discards it so we start clean rather
# than handing a poisoned connection back out.
engine.dispose()

tables_to_upload = {
    "dim_songs": dim_songs,
    "dim_dates": dim_dates,
    "fact_chart_entries": fact_chart_entries,
}

# Each table gets its OWN connection via engine.begin(), used once and
# closed immediately. This is the actual fix: last time, one failed
# statement on a shared/reused connection left that connection's
# transaction invalid, and every call after it inherited the same broken
# state. Isolating each table's upload to its own connection means one
# failure can't cascade into the next table.
for table_name, df in tables_to_upload.items():
    try:
        with engine.begin() as conn:
            df.to_sql(table_name, con=conn, if_exists="replace", index=False, chunksize=1000)
        print(f"{table_name}: uploaded, {df.shape[0]} rows")
    except Exception as e:
        print(f"{table_name}: FAILED during upload")
        print(traceback.format_exc())

print()

# Read-back verification, also on its own fresh connection per table.
for table_name, df in tables_to_upload.items():
    try:
        with engine.connect() as conn:
            check = pd.read_sql(f"SELECT COUNT(*) AS n FROM {table_name}", con=conn)
        actual = check["n"].iloc[0]
        expected = df.shape[0]
        status = "OK" if actual == expected else "MISMATCH"
        print(f"{table_name}: expected {expected}, got {actual} — {status}")
    except Exception as e:
        print(f"{table_name}: FAILED during verification")
        print(traceback.format_exc())

dim_songs: uploaded, 817 rows
dim_dates: uploaded, 560 rows
fact_chart_entries: uploaded, 27800 rows

dim_songs: expected 817, got 817 — OK
dim_dates: expected 560, got 560 — OK
fact_chart_entries: expected 27800, got 27800 — OK


### Checking for Hidden Duplicate `song_key` Values

Power BI's relationship builder rejected `dim_songs[song_key]` as the "one"
side of a relationship, reporting a duplicate value — even though the
song_key uniqueness check back when we built it (Recommendation 3) passed.

That's not a contradiction. That earlier check compared two *derived*
counts (distinct song_key values vs. distinct (song, artist) pairs) built
from the same source columns — if two rows differ by something invisible
like a trailing space or a different apostrophe character, both counts get
inflated by the same hidden near-duplicate, so the check silently passes.
It was never actually verifying against a true unique count.

This check re-fetches dim_songs directly from Supabase (not the local
kernel's cached copy) and uses repr() to reveal exactly what's stored,
including any hidden characters a normal print() would hide.

In [42]:
# Re-fetch dim_songs directly from Supabase to inspect exactly what's stored —
# not what's cached in the local kernel, in case anything changed since upload.
dim_songs_check = pd.read_sql("SELECT * FROM dim_songs", con=engine)

dupes = dim_songs_check[dim_songs_check.duplicated(subset="song_key", keep=False)].sort_values("song_key")
print(f"Duplicate song_key rows found: {len(dupes)}")

# repr() shows hidden characters (extra spaces, different quote marks, etc.)
# that a normal print() would render identically, hiding the real difference.
for idx, row in dupes.iterrows():
    print(repr(row["song_key"]))
    print("  song:  ", repr(row["song"]))
    print("  artist:", repr(row["artist"]))
    print()

Duplicate song_key rows found: 0


## Section 6: Summary and Next Steps

This notebook took spotify_top_50_world.csv — a single flat table, one row
per song per chart day — from raw upload through to a verified star-schema
model, following an ELT-style discovery process rather than assuming the
data was clean upfront.

**What EDA surfaced (Section 4):**
- Structurally sound: no missing values, no full duplicates, no duplicate
  (date, position) pairs, position/popularity both within expected bounds
- release_date stored at three different precisions (full date, year-month,
  year-only)
- song alone is not a safe identifier — 31 titles shared across artists
- One real artist-name encoding corruption (Ty Dolla $ign, 120 rows)
- One unresolved attribution inconsistency (MTG QUEM NAO QUER SOU EU)
- popularity == 0 in 182 rows, cause still unexplained
- 4 calendar days with zero snapshots (2024-03-14, 03-25, 07-11, 08-13),
  found only because dim_dates was built independently of the source data
  rather than derived from it

**What we built (Section 5):**
- release_date_cleaned — standardized, documented padding assumption
- Ty Dolla $ign artist name unified via exact-string replace
- song_key composite identifier (song + artist), since song alone collides
- Three normalized tables: dim_songs (820 rows), dim_dates (560 rows,
  gapless calendar), fact_chart_entries (27,800 rows, verified zero
  orphaned foreign keys against both dimensions)

**Uploaded to Supabase:** `dim_songs` (820 rows), `dim_dates` (560 rows), 
and `fact_chart_entries` (27,800 rows) — all three uploaded and 
independently verified via read-back row-count checks.

**Not yet resolved, carried forward as open questions:**
- Cause of popularity == 0 rows
- MTG QUEM NAO QUER SOU EU dual attribution

**Next steps:** move into Power BI — import all three tables, build the two
relationships (`dim_songs[song_key]` → `fact[song_key]`, 
`dim_dates[date]` → `fact[date]`), and mark `dim_dates` as an official Date 
Table before building out the DAX measure layer.

## Appendix: How to Use This Notebook

### What this notebook does

This notebook takes `spotify-top-50-world.csv` — a raw, flat, one-row-per-
song-per-chart-day dataset — from an untouched CSV all the way through to a
verified, normalized star-schema model living in a Supabase (Postgres)
database, ready to be plugged into Power BI. Along the way it:

- Loads and uploads the raw data with **zero cleaning** first (an ELT
  pattern — Extract, Load, *then* Transform), so there's always an
  untouched original to go back to
- Masks sensitive credentials on-screen, so the notebook is safe to run
  during a live demo, screen share, or recording
- Runs a full **discovery-based EDA** — missing values, duplicates,
  identifier safety, format consistency, encoding corruption, and a
  calendar-gap check — documenting exactly what was found and why it
  matters, not just what was fixed
- Applies each fix as an explicit, explained recommendation tied back to a
  specific finding, never a silent or unexplained change
- Splits the flat table into three normalized tables (`dim_songs`,
  `dim_dates`, `fact_chart_entries`) and **verifies** the model — zero
  orphaned foreign keys — before trusting it
- Uploads the final model back to Supabase with per-table connections, so
  one failed upload can't corrupt the others, and independently confirms
  every row count on the way back out
- Is **safe to re-run from top to bottom** any number of times — every
  `to_sql()` call uses `if_exists="replace"`, and every transformation is
  recomputed from source columns rather than layered on a previous run

### Where to find each section

| Section | What's there |
|---|---|
| **Section 1 — Loading the Data** | Prompts for your CSV path, loads it into `df` |
| **Section 2 — Connecting to Supabase** | Prompts for your credentials JSON path, builds and tests the database connection |
| **Section 3 — Uploading the Raw Data** | Uploads the untouched CSV as `spotify_top_50_world`, reads it back into `df_db` (the source of truth from here on) |
| **Section 4 — Exploring the Data** | The full EDA — missing values, duplicates, identifier checks, date-format inconsistency, encoding corruption, numeric range checks |
| **Section 4 Summary** | Consolidated list of every finding, plus a methodology note on avoiding false positives |
| **Section 5 — Recommendations** | Three fixes (`release_date` cleaning, artist-encoding fix, splitting into a star schema), each tied to a specific finding above |
| **Section 6 — Summary and Next Steps** | What was found, what was built, upload confirmation, and what comes next in Power BI |

### Setting this up in your own folder

This notebook doesn't hard-code any file paths — it asks for them at
runtime, so it runs identically no matter where you put it. To set it up:

1. **Create one project folder** anywhere on your machine, e.g.
   `Documents/Spotify_Project/`
2. **Place three files together in that folder:**
   - `Spotify_Project_EDA.ipynb` (this notebook)
   - `spotify-top-50-world.csv` (the raw dataset)
   - `supabase_credentials.json` (your own 5-key credentials file — see
     format below)
3. **Install the required libraries once**, before your first run:
   ```
   pip install pandas sqlalchemy psycopg2-binary
   ```
4. **Run the notebook.** When Section 1 and Section 2 prompt you for file
   paths, paste the **full path** to each file in your own folder, e.g.
   `/Users/yourname/Documents/Spotify_Project/spotify-top-50-world.csv`
   (Mac/Linux) or `C:\Users\yourname\Documents\Spotify_Project\spotify-
   top-50-world.csv` (Windows). Since nothing is hard-coded, this works
   the same whether the folder lives on your Desktop, in Documents, or
   anywhere else.

**`supabase_credentials.json` format** (5 keys, matching your own Supabase
project's Session Pooler connection details):
```json
{
  "db_host": "your-project-pooler-host",
  "db_port": 5432,
  "db_name": "postgres",
  "db_user": "your-pooler-username",
  "db_password": "your-current-password"
}
```

**Keep this file private.** If this folder is ever put under version
control (Git), add `supabase_credentials.json` to `.gitignore` immediately
— never commit it, and never paste its contents anywhere outside your own
